### PR TITLE
Introduce the bufr_query library from NOAA-EMC

### LINKED ISSUE #1299: Move the bufr-query patch to a patch manifest in the NOAA-EMC/bufr-query repo
The bufr-query recipe is in an open PR in the authoritative spack repo: https://github.com/spack/spack/pull/45920, and a request by the reviewer was made to move the patch to a "patch manifest" in the bufr-query repo at EMC.

I lost track of this request so I am putting this here so I can track the issue with ZenHub. I'm planning on getting to this in next week.

### PR DESCRIPTION
**Pull request by srherbener**
This PR adds in a new package.py script, plus a patch script, for the new bufr_query library from NOAA-EMC. This is being used by JEDI and other applications.

Note that I created this PR by cherry-picking the JCSDA/spack PR (https://github.com/JCSDA/spack/pull/461) merge commit.

Resolves https://github.com/JCSDA/spack-stack/issues/1299

### CODE REVIEW

**Review by climbfuji — APPROVED** [review]
This is identical to the well-tested addition in our fork. Thanks!

File: var/spack/repos/builtin/packages/bufr-query/package.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    depends_on("netcdf-c", type=("build", "run"))
+    depends_on("netcdf-cxx4", type=("build", "run"))
+    depends_on("bufr", type=("build", "run"))
+
+    # Optional dependencies
+    variant("python", default=True, description="Enable Python interface")
+
+    with when("+python"):
+        extends("python")
+        depends_on("py-pybind11")
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/bufr-query/package.py]: Is this sufficient as a build dependency?
```suggestion
        depends_on("py-pybind11", type="build")
```

File: var/spack/repos/builtin/packages/bufr-query/package.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    # Patches
+    patch("site-packages.patch", when="+python @:0.0.2")
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/bufr-query/package.py]: Can we use https://github.com/NOAA-EMC/bufr-query/pull/20.patch?full_index=1 here to make origin of patch manifest?
```suggestion
    patch(
        "https://github.com/NOAA-EMC/bufr-query/pull/20.patch?full_index=1",
        sha256="",
        when="+python @:0.0.2",
    )
```

**Review by wdconinc — COMMENTED** [review]
Thanks for the new package contribution. Two comments (and I guess I see 0.0.3 is available so maybe you want to add it).

File: var/spack/repos/builtin/packages/bufr-query/package.py (hunk 1 above)
Comment (srherbener) [reply var/spack/repos/builtin/packages/bufr-query/package.py]: Sorry about the delay. Thank you for the suggested code. I agree that this is a good improvement.

File: var/spack/repos/builtin/packages/bufr-query/package.py (hunk 2 above)
Comment (srherbener) [reply var/spack/repos/builtin/packages/bufr-query/package.py]: Thanks for the suggested code. I think this is a good improvement! How do I get the sha256 value. Can I get that from the existing patch file? Or do I commit these and get the sha256 value from an error/warning message? Thanks!

Comment (wdconinc) [reply var/spack/repos/builtin/packages/bufr-query/package.py]: I usually do something like
```
$ curl -L https://github.com/NOAA-EMC/bufr-query/pull/20.patch?full_index=1 | sha256sum
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  5177    0  5177    0     0  11118      0 --:--:-- --:--:-- --:--:-- 11118
3acf11082c9e76e64dbbda4f62ac0cbc234dca7e60c85a275e778417cfd65001  -
```
which then becomes:
```suggestion
    patch(
        "https://github.com/NOAA-EMC/bufr-query/pull/20.patch?full_index=1",
        sha256="3acf11082c9e76e64dbbda4f62ac0cbc234dca7e60c85a275e778417cfd65001",
        when="+python @:0.0.2",
    )
```

Comment (srherbener) [reply var/spack/repos/builtin/packages/bufr-query/package.py]: Great! Thank you for showing me how to get the sha256 value and setting up the suggested code change! (I get the same result when I run your curl example on my Mac.)

Comment (srherbener) [reply var/spack/repos/builtin/packages/bufr-query/package.py]: I still need to remove the `site-packages.patch` file, correct?

Comment (wdconinc) [reply var/spack/repos/builtin/packages/bufr-query/package.py]: Yes, that's not necessary anymore now.

**Comment by srherbener** [comment]
I removed the site-packages.patch file and added the new (0.0.3) version.

**Review by wdconinc — APPROVED** [review]
Looks good. Checksum for 0.0.3 verified as well.

**Pull request merged**